src/add-hooks.coffee

```diff
@@ -101,30 +101,38 @@ addHooks = (runner, transactions, callback) ->
   else
     # Expand hookfiles - sort files alphabetically and resolve their paths
     hookfiles = [].concat runner.configuration?.options?.hookfiles
-    files = hookfiles.reduce((result, unresolvedPath) ->
-      # glob.sync does not resolve paths, only glob patterns
-      unresolvedPaths = if glob.hasMagic(unresolvedPath) then glob.sync(unresolvedPath) else [unresolvedPath]
-
-      # Gradually append sorted and resolved paths
-      result.concat unresolvedPaths
-        # Create a filename / filepath map for easier sorting
-        # Example:
-        # [
-        #   { basename: 'filename1.coffee', path: './path/to/filename1.coffee' }
-        #   { basename: 'filename2.coffee', path: './path/to/filename2.coffee' }
-        # ]
-        .map((filepath) -> basename: basename(filepath), path: filepath)
-        # Sort 'em up
-        .sort((a, b) -> switch
-          when a.basename < b.basename then -1
-          when a.basename > b.basename then 1
-          else 0
-        )
-        # Resolve paths to absolute form. Take into account user defined current
-        # working directory, fallback to process.cwd() otherwise
-        .map((item) -> path.resolve(customConfigCwd or process.cwd(), item.path))
-    , [] # Start with empty result
-    )
+    try
+      files = hookfiles.reduce((result, unresolvedPath) ->
+        # glob.sync does not resolve paths, only glob patterns
+
+        unresolvedPaths = if glob.hasMagic(unresolvedPath) then glob.sync(unresolvedPath) else
+          if fs.existsSync(unresolvedPath) then [unresolvedPath] else []
+
+        if unresolvedPaths.length == 0
+          throw new Error("Hook file(s) not found on path: #{unresolvedPath}")
```

Comment:
I don't like the fact that `throw new Error(...)` is used inside a `reduce` method. I would rather seen it rewritten in `forEach` or some imperative equivalent.

Reply:
I added some commits. Not sure if I actually addressed this comment though, because the resulting flow avoids reduce, but now the `throw` still happens in a `.map()` iteratee.

I think throwing exceptions isn't anything a pure function couldn't do. It's not mutation, it's just one of the possible outcomes of the pure function.